jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildTarballAndLoadDockerStep.java

```diff
@@ -0,0 +1,131 @@
+      ListenableFuture<List<ListenableFuture<CachedLayer>>> pullBaseImageLayerFuturesFuture,
+      List<ListenableFuture<CachedLayer>> buildApplicationLayerFutures,
+      ListenableFuture<ListenableFuture<Blob>> buildConfigurationFutureFuture) {
+    this.buildConfiguration = buildConfiguration;
+    this.listeningExecutorService = listeningExecutorService;
+    this.pullBaseImageLayerFuturesFuture = pullBaseImageLayerFuturesFuture;
+    this.buildApplicationLayerFutures = buildApplicationLayerFutures;
+    this.buildConfigurationFutureFuture = buildConfigurationFutureFuture;
+  }
+
+  /**
+   * Depends on {@code pullBaseImageLayerFuturesFuture} and {@code buildConfigurationFutureFuture}.
+   */
+  @Override
+  public Void call() throws ExecutionException, InterruptedException {
+    List<ListenableFuture<?>> dependencies = new ArrayList<>();
+    dependencies.addAll(NonBlockingFutures.get(pullBaseImageLayerFuturesFuture));
+    dependencies.addAll(buildApplicationLayerFutures);
+    dependencies.add(NonBlockingFutures.get(buildConfigurationFutureFuture));
+    return Futures.whenAllComplete(dependencies)
+        .call(this::afterPushBaseImageLayerFuturesFuture, listeningExecutorService)
+        .get();
+  }
+
+  /**
+   * Depends on {@code pushAuthorizationFuture}, {@code pushBaseImageLayerFuturesFuture.get()},
+   * {@code pushApplicationLayerFutures}, and (@code
+   * containerConfigurationBlobDescriptorFutureFuture.get()}.
+   *
+   * <p>TODO: Refactor into testable components
+   */
+  private Void afterPushBaseImageLayerFuturesFuture()
+      throws ExecutionException, InterruptedException, IOException {
+    // Add layers to image tarball
+    TarStreamBuilder tarStreamBuilder = new TarStreamBuilder();
+    List<String> layerFiles = new ArrayList<>();
+    for (Future<CachedLayer> cachedLayerFuture :
+        NonBlockingFutures.get(pullBaseImageLayerFuturesFuture)) {
+      Path layerFile = NonBlockingFutures.get(cachedLayerFuture).getContentFile();
+      layerFiles.add(layerFile.getFileName().toString());
+      tarStreamBuilder.addEntry(
+          new TarArchiveEntry(layerFile.toFile(), layerFile.getFileName().toString()));
+    }
+    for (Future<CachedLayer> cachedLayerFuture : buildApplicationLayerFutures) {
+      Path layerFile = NonBlockingFutures.get(cachedLayerFuture).getContentFile();
+      layerFiles.add(layerFile.getFileName().toString());
+      tarStreamBuilder.addEntry(
+          new TarArchiveEntry(layerFile.toFile(), layerFile.getFileName().toString()));
+    }
+
+    // Add config to tarball
+    // TODO: Add ability to add blobs as entries to TarStreamBuilder without using temp files
+    Blob containerConfigurationBlob =
+        NonBlockingFutures.get(NonBlockingFutures.get(buildConfigurationFutureFuture));
+    Path tempConfig = Files.createTempFile(null, null);
+    tempConfig.toFile().deleteOnExit();
+    try (OutputStream bufferedOutputStream = Files.newOutputStream(tempConfig)) {
```

Comment:
Looks like this is missing a `new BufferedOutputStream`. Although without it is fine, buffering the outputstream should make it be faster for writing to IO by reducing the total amount of expensive IO operations.

Reply:
Yeah my b, I looked this up and misinterpreted the results I found, I'll change it back.